src/io/hdf_hyperslab.h

```diff
@@ -12,202 +12,164 @@
+                << sizes_file[dim] << " mismatched!" << std::endl;
+        throw std::runtime_error(err_msg.str());
       }
-      return h5d_read(grp, aname, ref_.data(), xfer_plist);
+
+      if (selected_space.dims[dim] == 0)
+        selected_space.dims[dim] = file_space.dims[dim];
+      total_size *= selected_space.dims[dim];
     }
-  }
-  inline bool write(hid_t grp, const std::string& aname, hid_t xfer_plist = H5P_DEFAULT)
-  {
-    std::cerr << " Disabled hyperslab write with boost::multi::array<gpu_allocator>.\n";
-    return false;
+
+    if (total_size != container_traits<CT>::getSize(ref_))
+      container_traits<CT>::resize(ref_, selected_space.dims, slab_rank);
   }
 };
 
-template<typename T, unsigned MAXDIM>
-struct h5data_proxy<hyperslab_proxy<boost::multi::array_ref<T, 2, qmc_cuda::cuda_gpu_ptr<T>>, MAXDIM>>
+template<typename CT, unsigned RANK>
+struct h5data_proxy<hyperslab_proxy<CT, RANK>>
 {
-  typedef boost::multi::array_ref<T, 2, qmc_cuda::cuda_gpu_ptr<T>> CT;
-  hyperslab_proxy<CT, MAXDIM>& ref_;
-  h5data_proxy(hyperslab_proxy<CT, MAXDIM>& a) : ref_(a) {}
+  hyperslab_proxy<CT, RANK>& ref_;
+
+  h5data_proxy(hyperslab_proxy<CT, RANK>& a) : ref_(a) {}
+
   inline bool read(hid_t grp, const std::string& aname, hid_t xfer_plist = H5P_DEFAULT)
   {
-    if (ref_.use_slab)
-    {
-      // later on specialize h5d_read for fancy pointers
-      std::size_t sz = ref_.ref.num_elements();
-      boost::multi::array<T, 1> buf(typename boost::multi::layout_t<1u>::extensions_type{sz});
-      auto ret = h5d_read(grp, aname.c_str(), ref_.slab_rank, ref_.slab_dims.data(), ref_.slab_dims_local.data(),
-                          ref_.slab_offset.data(), buf.origin(), xfer_plist);
-      qmc_cuda::copy_n(buf.data(), sz, ref_.ref.origin());
-      return ret;
-    }
-    else
-    {
-      int rank = ref_.slab_rank;
-      if (!get_space(grp, aname, rank, ref_.slab_dims.data(), true))
-      {
-        std::cerr << " Disabled hyperslab resize with boost::multi::array_ref<gpu_ptr>.\n";
-        return false;
-      }
-      return h5d_read(grp, aname, ref_.data(), xfer_plist);
-    }
+    std::vector<hsize_t> sizes_file;
+    getDataShape<typename hyperslab_proxy<CT, RANK>::element_type>(grp, aname, sizes_file);
+    ref_.adaptShape(sizes_file);
+    ref_.checkUserRankSizes();
+    return h5d_read(grp, aname.c_str(), ref_.file_space.rank, ref_.file_space.dims, ref_.selected_space.dims,
+                    ref_.slab_offset.data(), hyperslab_proxy<CT, RANK>::SpaceType::get_address(container_traits<CT>::getElementPtr(ref_.ref_)),
+                    xfer_plist);
   }
+
   inline bool write(hid_t grp, const std::string& aname, hid_t xfer_plist = H5P_DEFAULT)
   {
-    std::cerr << " Disabled hyperslab write with boost::multi::array_ref<gpu_ptr>.\n";
-    return false;
+    ref_.checkUserRankSizes();
+    return h5d_write(grp, aname.c_str(), ref_.file_space.rank, ref_.file_space.dims, ref_.selected_space.dims,
```

Comment:
@lshulen you are right. I need to check size before writing to see if sufficient size is provided in the container.

Reply:
If the type supports that check, then I think it would be a good thing.

Reply:
checkContainerCapacity now handles container size checking.